system-addon/lib/DownloadsManager.jsm

```diff
@@ -0,0 +1,183 @@
+    this._store = null;
+    this._viewableDownloadItems = new Map();
+    this._downloadTimer = null;
+  }
+
+  setTimeout(callback, delay) {
+    let timer = Cc["@mozilla.org/timer;1"].createInstance(Ci.nsITimer);
+    timer.initWithCallback(callback, delay, Ci.nsITimer.TYPE_ONE_SHOT);
+    return timer;
+  }
+
+  formatDownload(element) {
+    const downloadedItem = element.download;
+    let description;
+    if (element.fileType) {
+      // If we have a file type: '1.5 MB — PNG'
+      description = `${element.sizeStrings.stateLabel} \u2014 ${element.fileType}`;
+    } else {
+      // If we do not have a file type: '1.5 MB'
+      description = `${element.sizeStrings.stateLabel}`;
+    }
+    return {
+      hostname: new URL(downloadedItem.source.url).hostname,
+      url: downloadedItem.source.url,
+      path: downloadedItem.target.path,
+      title: element.displayName,
+      description,
+      referrer: downloadedItem.source.referrer
+    };
+  }
+
+  init(store) {
+    this._store = store;
+    this._browser = Services.appShell.hiddenDOMWindow;
+    this._downloadData = DownloadsCommon.getData(this._browser.ownerGlobal, true, false, true);
+    this._downloadData.addView(this);
+  }
+
+  onDownloadAdded(download) {
+    const elem = new DownloadElement(download, this._browser);
+    const downloadedItem = elem.download;
+    if (!this._viewableDownloadItems.has(downloadedItem.source.url)) {
+      this._viewableDownloadItems.set(downloadedItem.source.url, elem);
+
+      // Debounce incoming onDownloadAdded notifications
+      if (this._downloadTimer) {
+        this._downloadTimer.delay = DOWNLOAD_CHANGED_DELAY_TIME;
+      } else {
+        this._downloadTimer = this.setTimeout(() => {
+          this._downloadTimer = null;
+          this._store.dispatch({type: at.DOWNLOAD_CHANGED});
+        }, DOWNLOAD_CHANGED_DELAY_TIME);
+      }
+    }
+  }
+
+  onDownloadRemoved(download) {
+    if (this._viewableDownloadItems.has(download.source.url)) {
+      this._viewableDownloadItems.delete(download.source.url);
+      this._store.dispatch({type: at.DOWNLOAD_CHANGED});
+    }
+  }
+
+  async getDownloads({threshold, numItems = this._viewableDownloadItems.size, onlyExists = false}) {
+    if (!threshold) {
+      return [];
+    }
+    let results = [];
+
+    // Only get downloads within the time threshold specified and sort by recency
+    let downloads = [...this._viewableDownloadItems.values()]
+                      .filter(elem => elem.download.endTime > Date.now() - threshold)
+                      .sort((elem1, elem2) => elem1.download.endTime < elem2.download.endTime);
+
+    // Iterate over this list checking for ones that exist, and formatting the ones
+    // to return
+    for (const elem of downloads) {
+      if (onlyExists) {
+        // Refresh download to ensure the 'exists' attribute is up to date
+        await elem.download.refresh();
+        if (elem.download.target.exists && elem.download.succeeded) {
```

Comment:
you can `if (!exists) continue;` and share the `formatDownload` / `else` stuff. Both paths want only succeeded downloads too?

Reply:
maybe pass `succeeded` as a parameter too?

Reply:
I suppose an optional succeeded parameter would match and doesn't need to add an extra branch for test coverage